src/Components/Components/src/Forms/EditContextDataAnnotationsExtensions.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    /// <summary>
+    /// Extension methods to add DataAnnotations validation to an <see cref="EditContext"/>.
+    /// </summary>
+    public static class EditContextDataAnnotationsExtensions
+    {
+        private static ConcurrentDictionary<(Type ModelType, string FieldName), PropertyInfo> _propertyInfoCache
+            = new ConcurrentDictionary<(Type, string), PropertyInfo>();
+
+        /// <summary>
+        /// Adds DataAnnotations validation support to the <see cref="EditContext"/>.
+        /// </summary>
+        /// <param name="editContext">The <see cref="EditContext"/>.</param>
+        public static EditContext AddDataAnnotationsValidation(this EditContext editContext)
+        {
+            if (editContext == null)
+            {
+                throw new ArgumentNullException(nameof(editContext));
+            }
+
+            var messages = new ValidationMessageStore(editContext);
+
+            // Perform object-level validation on request
+            editContext.OnValidationRequested +=
+                (sender, eventArgs) => ValidateModel((EditContext)sender, messages);
+
+            // Perform per-field validation on each field edit
+            editContext.OnFieldChanged +=
+                (sender, fieldIdentifier) => ValidateField(editContext, messages, fieldIdentifier);
+
+            return editContext;
```

Comment:
This is a very non-C# way of wiring something up, and it's also really abstraction-light which is probably fine.

The only concern that I have about this is that if these events have mixed use (used by validation providers, used by components/apps) that becomes really hard to reason about due to ordering. I don't have any really big concerns about this at this point. I think all of the other approaches to how something like data annotations gets wired up have their own drawbacks as well.

Reply:
Agreed. I *hope* people stick within the one-way flow of update propagation, because then ordering is largely not a concern. But I recognize people might do really dubious things like having a validation mechanism that varies its behavior based on validation messages registered by a different validation system.

I've never really believed in the idea of solving ordering issues by having an `Order` integer - it only scales so far, before you can't manage it either (e.g., validation systems registered by multiple independent NuGet packages that need to be ordered). `Order` only works internally within a single, sufficiently small codebase.

Let's see how it goes.